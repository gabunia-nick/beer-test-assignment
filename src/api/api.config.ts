export const API_BASE_URL = 'https://api.punkapi.com/v2/';

export const API_RANDOM_BEER = 'beers/random';
