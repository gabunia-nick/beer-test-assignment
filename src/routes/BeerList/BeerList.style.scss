.BeerList {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  padding: 10px;

  .BeerCard {
    margin-bottom: 15px;
    width: 100%;

    @include desktop {
      width: calc(50% - 15px);
    }
  }
}
