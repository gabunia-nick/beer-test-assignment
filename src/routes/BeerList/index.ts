export { default } from './BeerList.component';
