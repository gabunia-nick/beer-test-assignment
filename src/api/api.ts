import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import { API_BASE_URL, API_RANDOM_BEER } from './api.config';
import { BeerResponse } from './types/BeerResponse.type';

export const beerApi = createApi({
  reducerPath: 'beerReducer',
  baseQuery: fetchBaseQuery({ baseUrl: API_BASE_URL }),
  endpoints: (builder) => ({
    getRandomBeers: builder.query<BeerResponse[], number>({
      async queryFn(beerNumber, _api, _extraOptions, fetchWithBQ) {
        const fetchBeersRecc = async (number: number, resArr: BeerResponse[], set: Set<number>)
        : Promise<BeerResponse[]> => {
          if (number <= 0) {
            return resArr;
          }

          const requestPromiseArray = Array(number).fill(null)
            .map(() => fetchWithBQ(API_RANDOM_BEER));
          const result = await Promise.all(requestPromiseArray);

          const newResArr = result.reduce((acc, curr) => {
            const { data } = curr;
            const [beer] = data as BeerResponse[] || [];

            if (!beer) {
              return acc;
            }

            if (set.has(beer.id)) {
              return acc;
            }

            set.add(beer.id);
            return [...acc, beer];
          }, resArr);

          return fetchBeersRecc(number - newResArr.length, newResArr, set);
        };

        const beerSet = new Set<number>();

        return { data: await fetchBeersRecc(beerNumber, [], beerSet) };
      },
    }),
  }),
});

export const { useGetRandomBeersQuery } = beerApi;
