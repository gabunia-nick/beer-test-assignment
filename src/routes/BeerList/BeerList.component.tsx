import React from 'react';
import { useGetRandomBeersQuery } from '../../api/api';
import BeerCard from '../../components/BeerCard';

import './BeerList.style.scss';

const BeerListComponent = () => {
  const { data, error, isLoading } = useGetRandomBeersQuery(30);

  return (
    <main className="BeerList">
      {isLoading && <p className="BeerList-LoadingMessage">Loading...</p>}
      {error && <p className="BeerList-ErrorMessage">Oops... Something went wrong</p>}
      {(!data && !error && !isLoading) && <p className="BeerList-NoResults">Unfortunately no beers were found</p>}
      {data?.map((beer) => (
        <BeerCard key={beer.id} name={beer.name} tagline={beer.tagline} imageUrl={beer.image_url} />
      ))}
    </main>
  );
};

export default BeerListComponent;
