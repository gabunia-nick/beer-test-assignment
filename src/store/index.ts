import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { beerApi } from '../api/api';

const store = configureStore({
  reducer: {
    [beerApi.reducerPath]: beerApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(beerApi.middleware),
});

setupListeners(store.dispatch);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;

export default store;
