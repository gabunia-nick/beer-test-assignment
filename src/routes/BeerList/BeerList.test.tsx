import React from 'react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { prettyDOM, render, screen, wait } from '@testing-library/react';
import '@testing-library/jest-dom';
import { API_BASE_URL, API_RANDOM_BEER } from '../../api/api.config';
import { Provider } from 'react-redux';
import store from '../../store';
import BeerList from '.';

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

const server = setupServer(
  rest.get(`${API_BASE_URL}${API_RANDOM_BEER}`, (req, res, ctx) => {
    return res(ctx.json([
        {
          id: Math.floor(Math.random() * 1000000),
          name: 'Imperial Stout',
          tagline: 'The most bitter beer you can eve drink',
          image_url: 'https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg'
        }
    ]))
  })
);

test('BeerList content is being rendered correctly', async () => {
  render(
    <Provider store={store}>
      <BeerList />
    </Provider>
  );

  await wait(() => screen.getAllByTestId('beercard-element'));

  const allCards = screen.getAllByTestId('beercard-element');

  expect(allCards.length).toBe(30);
  expect(prettyDOM(screen.getByTestId('beerlist-element'))).toMatchSnapshot();
});
