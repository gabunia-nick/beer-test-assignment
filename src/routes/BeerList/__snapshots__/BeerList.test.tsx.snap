// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerList content is being rendered correctly 1`] = `
"[36m<main[39m
  [33mclass[39m=[32m\\"BeerList\\"[39m
  [33mdata-testid[39m=[32m\\"beerlist-element\\"[39m
[36m>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-Content\\"[39m
    [36m>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Tagline\\"[39m
      [36m>[39m
        [0mThe most bitter beer you can eve drink[0m
      [36m</p>[39m
      [36m<p[39m
        [33mclass[39m=[32m\\"BeerCard-Name\\"[39m
      [36m>[39m
        [0mImperial Stout[0m
      [36m</p>[39m
    [36m</div>[39m
  [36m</div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"BeerCard\\"[39m
    [33mdata-testid[39m=[32m\\"beercard-element\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"BeerCard-ImageContainer\\"[39m
    [36m>[39m
      [36m<img[39m
        [33malt[39m=[32m\\"Imperial Stout\\"[39m
        [33mclass[39m=[32m\\"BeerCard-Image\\"[39m
        [33msrc[39m=[32m\\"https://image.shutterstock.com/image-photo/beer-600w-617336486.jpg\\"[39m
      [36m/>[39m
    [36m</div>[39m
    [3..."
`;
